docs/proposals/nginx-extensions.md

```diff
@@ -0,0 +1,964 @@
+
+_Extension point:_ HTTPRoute, HTTPRoute rule
+
+_NGINX context(s):_ http, server, location
+
+Features:
+
+- OTEL Tracing: enable/disable by Route or routing rule. Set span name, attributes, and context.
+
+NGINX directives:
+
+- [`otel_trace`](https://nginx.org/en/docs/ngx_otel_module.html#otel_trace)
+- [`otel_trace_context`](https://nginx.org/en/docs/ngx_otel_module.html#otel_trace_context)
+- [`otel_span_name`](https://nginx.org/en/docs/ngx_otel_module.html#otel_span_name)
+- [`otel_span_attr`](https://nginx.org/en/docs/ngx_otel_module.html#otel_span_attr)
+
+#### Future Work
+
+Add support for:
+
+- OTEL logging
+- Metrics
+
+#### Alternatives
+
+- Combine with OTEL settings in Gateway Settings for one OTEL Policy: Rather than splitting tracing across two Policies, we could create a single tracing Policy. The issue with this approach is that some tracing settings -- such as exporter endpoint -- should be restricted to Cluster Operators, while settings like attributes should be available to Application Developers. If we combine these settings, RBAC will not be sufficient to restrict access across the settings. We will have to disallow certain fields based on the resource the Policy is attached to. This is a bad user experience.
+- Inherited Policy: An Inherited Policy would be useful if there is a use case for the Cluster Operator enforcing or defaulting the OTEL tracing settings included in this policy.
+
+### Proxy Settings
+
+_Extension type:_ Inherited Policy
+
+_Resource type:_ CRD
+
+_Role(s):_ Cluster Operator, Application Developer
+
+_Extension point:_ Gateway, HTTPRoute, HTTPRoute Rule
+
+_NGINX context(s):_ http, server, location
+
+Features:
+
+- Proxy buffering
+- Connection timeouts
+- Next upstream retries
+
+NGINX directives:
+
+- [`proxy_connect_timeout`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_connect_timeout)
+- [`proxy_read_timeout`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout)
+- [`proxy_send_timeout`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_send_timeout)
+- [`proxy_next_upstream`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_next_upstream)
+- [`proxy_next_upstream_timeout`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_next_upstream_timeout)
+- [`proxy_next_upstream_retries`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_next_upstream_tries)
+- [`proxy_buffering`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_buffering)
+
+#### Future Work
+
+- Add other `proxy_*` directives
+
+#### Alternatives
+
+- Direct Policy: If there's no strong use case for the Cluster Operator setting sane defaults for these settings, then we can use a Direct Policy. The Direct Policy could attach to an HTTPRoute or HTTPRoute Rule, and the NGINX contexts would be server and location.
+
+### Circuit Breaker/ Backup service
+
+_Extension type:_ Direct Policy/Filter
```

Comment:
What if we have a custom BackendRef for that:

```yaml
apiVersion: . . .
kind: CircuitBreaker
metadata:
  name: circuit-breaker-example
spec:
  primary:
    service: coffee-primary
    port: 80
  fallback:
    service: coffee-backup
    port: 80
```

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: custom-backend
spec:
  hostnames:
  - "example.com"
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /
    backendRefs:
    - group: my.group.io
      kind: CircuitBreaker
      name: circuit-breaker-example
      port: 80
```

Reply:
That's a really cool idea. I will add it to the possibilities.